### PR TITLE
TW-700: Separated pinned chat with category in chat list

### PR DESCRIPTION
**Pull request by nqhhdev**
### Ticket:

- #700 

### Resolved

https://github.com/linagora/twake-on-matrix/assets/99852347/6ffbcf0d-ddca-4f80-83c3-e1c56ebddb97

#### Share file 

https://github.com/linagora/twake-on-matrix/assets/99852347/6de62321-00c8-4bbf-9055-eb2d3076adff

#### Forward

https://github.com/linagora/twake-on-matrix/assets/99852347/bc52edca-fb66-4450-9e19-6ba871231909

### CODE REVIEW

File: lib/pages/chat_list/chat_list.dart [hunk 1]
```diff
@@ -67,123 +65,100 @@ class ChatListController extends State<ChatList>
...
+  bool get displayNavigationBar => false;
 
-  bool isSearching = false;
-  static const String _serverStoreNamespace = 'im.fluffychat.search.server';
+  Stream<Client> get clientStream => _clientStream.stream;
 
-  final TextEditingController searchChatController = TextEditingController();
+  // Needs to match GroupsSpacesEntry for 'separate group' checking.
+  List<Room> get spaces =>
+      Matrix.of(context).client.rooms.where((r) => r.isSpace).toList();
```
Comment (imGok) [inline lib/pages/chat_list/chat_list.dart:121]:
```suggestion
  List<Room> get spaces =>
      client.rooms.where((r) => r.isSpace).toList();
```

File: lib/pages/chat_list/chat_list_body_view_style.dart [hunk 2]
```diff
@@ -0,0 +1,37 @@
+import 'package:flutter/cupertino.dart';
+
+class ChatListBodyViewStyle {
+  static const double sizeIconExpand = 24;
+
+  static double heightIsTorBrowser(bool isTorBrowser) => isTorBrowser ? 64 : 0;
```
Comment (imGok) [inline lib/pages/chat_list/chat_list_body_view_style.dart:6]: imo it should be
```suggestion
  static double torContainerHeight(bool isTorBrowser) => isTorBrowser ? 64 : 0;
```

File: lib/pages/chat_list/chat_list_item_style.dart [hunk 3]
```diff
@@ -3,9 +3,17 @@ import 'package:linagora_design_flutter/colors/linagora_ref_colors.dart';
...
+
+  static const double readIconSize = 20;
+
+  static const double groupIconSize = 16;
+
+  static const double mentionIconWidth = 20;
+
+  static const Color readMessageColor = Color(0xFF787579);
+
+  static const Color unreadMessageOfGroupColor = Color(0xFF1C1B1F);
```
Comment (hoangdat) [inline lib/pages/chat_list/chat_list_item_style.dart:15]: no color from LinagoraSystem?

**Review by hoangdat — APPROVED** [review]

File: lib/pages/chat_list/chat_list_item_style.dart (hunk 3 above)
Comment (nqhhdev) [reply lib/pages/chat_list/chat_list_item_style.dart:15]: Updated

Screenshot 2023-10-06 at 13 40 32

File: lib/pages/chat_list/chat_list_item_style.dart [hunk 4]
```diff
@@ -3,9 +3,15 @@ import 'package:linagora_design_flutter/colors/linagora_ref_colors.dart';
...
-  static double get mentionIconWidth => 20;
-  static Color get readMessageColor => const Color(0xFF787579);
+
+  static const double readIconSize = 20;
+
+  static const double groupIconSize = 16;
+
+  static const double mentionIconWidth = 20;
+
+  static const Color readMessageColor = Color(0xFF787579);
```
Comment (hoangdat) [inline lib/pages/chat_list/chat_list_item_style.dart:13]: wrong color

File: lib/pages/chat_list/chat_list_item.dart [hunk 5]
```diff
@@ -84,9 +85,14 @@ class ChatListItem extends StatelessWidget with ChatListItemMixin {
     }
   }
 
+  bool get _isGroupChat => !room.isDirectChat;
+
+  bool get _isMaskAsUnread =>
+      room.isUnread || room.membership == Membership.invite;
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list_item.dart:91]: why membership invite is related in unread

File: lib/pages/chat_list/chat_list_item.dart [hunk 6]
```diff
@@ -84,9 +85,14 @@ class ChatListItem extends StatelessWidget with ChatListItemMixin {
     }
   }
 
+  bool get _isGroupChat => !room.isDirectChat;
+
+  bool get _isMaskAsUnread =>
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list_item.dart:90]:
```suggestion
  bool get _isMarkAsUnread =>
```

File: lib/pages/chat_list/chat_list_item.dart [hunk 7]
```diff
@@ -110,10 +116,34 @@ class ChatListItem extends StatelessWidget with ChatListItemMixin {
...
+                      ),
+                      if (_isGroupChat)
+                        Positioned(
+                          bottom: 0,
+                          right: 0,
+                          child: Container(
+                            padding: ChatListItemStyle.paddingIconGroup,
+                            decoration: const BoxDecoration(
+                              shape: BoxShape.circle,
+                              color: Colors.white,
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list_item.dart:134]: please move it to style

File: lib/pages/chat_list/chat_list_view_builder.dart [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+  @override
+  Widget build(BuildContext context) {
+    return ListView.builder(
+      shrinkWrap: true,
+      physics: const NeverScrollableScrollPhysics(),
+      itemCount: rooms.length,
+      itemBuilder: (BuildContext context, int index) {
+        return ValueListenableBuilder(
+          valueListenable: controller.selectModeNotifier,
+          builder: (context, selectedRoomIds, _) {
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list_view_builder.dart:26]: i don't see where you use `selectedRoomIds`

**Review by drminh2807 — APPROVED** [review]

File: lib/pages/chat_list/chat_list_view_builder.dart (hunk 8 above)
Comment (nqhhdev) [reply lib/pages/chat_list/chat_list_view_builder.dart:26]: Yes, it triggers rebuild when `selectMode` change

File: lib/pages/chat_list/chat_list_body_view.dart [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+                },
+              ),
+            ),
+          ),
+        ),
+      ],
+    );
+  }
+
+  Widget _expandableTitleBuilder({
```
Comment (hoangdat) [inline lib/pages/chat_list/chat_list_body_view.dart:230]: how about a stateless?

File: lib/presentation/decorators/chat_list/subtitle_text_style_decorator/subtitle_text_style_view.dart [hunk 10]
```diff
@@ -0,0 +1,12 @@
+import 'package:fluffychat/presentation/decorators/chat_list/subtitle_text_style_decorator/subtitle_text_style_decorator.dart';
+
+class ChatLitSubSubtitleTextStyleView {
+  static final SubtitleReadTextStyleDecorator _subtitleReadTextStyleDecorator =
+      SubtitleReadTextStyleDecorator();
+  static final SubtitleUnreadTextStyleDecorator _unreadTextStyleDecorator =
+      SubtitleUnreadTextStyleDecorator(_subtitleReadTextStyleDecorator);
+  static final SubtitleMuteTextStyleDecorator _muteTextStyleDecorator =
+      SubtitleMuteTextStyleDecorator(_unreadTextStyleDecorator);
```
Comment (hoangdat) [inline lib/presentation/decorators/chat_list/subtitle_text_style_decorator/subtitle_text_style_view.dart:9]: IMO, no need assignment. WDYT?

File: lib/presentation/decorators/chat_list/title_text_style_decorator/title_text_style_component.dart [hunk 11]
```diff
@@ -0,0 +1,6 @@
+import 'package:flutter/material.dart';
+import 'package:matrix/matrix.dart';
+
+abstract class InterfaceChatListTitleTextStyleComponent {
```
Comment (hoangdat) [inline lib/presentation/decorators/chat_list/title_text_style_decorator/title_text_style_component.dart:4]:
```suggestion
abstract class IChatListTitleTextStyleComponent {
```

Comment (hoangdat) [inline lib/presentation/decorators/chat_list/title_text_style_decorator/title_text_style_component.dart:4]:
```suggestion
abstract class ChatListTitleTextStyleComponent {
```

File: lib/presentation/decorators/chat_list/title_text_style_decorator/title_text_style_decorator.dart [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+  TextStyle textStyle(Room room) {
+    return _interfaceTextStyleComponent.textStyle(room);
+  }
+
+  @override
+  InterfaceChatListTitleTextStyleComponent get interfaceTextStyleComponent =>
+      _interfaceTextStyleComponent;
+}
+
+class TitleReadTextStyleDecorator
```
Comment (hoangdat) [inline lib/presentation/decorators/chat_list/title_text_style_decorator/title_text_style_decorator.dart:28]:
```suggestion
class ReadChatListTitleTextStyleDecorator
```

File: lib/presentation/decorators/chat_list/title_text_style_decorator/title_text_style_decorator.dart [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    implements InterfaceChatListTitleTextStyleComponent {
+  @override
+  TextStyle textStyle(Room room) {
+    return LinagoraTextStyle.material().bodyLarge2.copyWith(
+          color: LinagoraSysColors.material().onSurface,
+        );
+  }
+}
+
+class TitleUnreadTextStyleDecorator implements TextStyleDecorator {
```
Comment (hoangdat) [inline lib/presentation/decorators/chat_list/title_text_style_decorator/title_text_style_decorator.dart:38]:
```suggestion
class UnreadChatListTitleTextStyleDecorator implements TextStyleDecorator {
```

File: lib/presentation/decorators/chat_list/title_text_style_decorator/title_text_style_decorator.dart [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+      return _interfaceTextStyleComponent.textStyle(room);
+    }
+  }
+
+  @override
+  InterfaceChatListTitleTextStyleComponent get interfaceTextStyleComponent =>
+      _interfaceTextStyleComponent;
+}
+
+class TitleMuteTextStyleDecorator implements TextStyleDecorator {
```
Comment (hoangdat) [inline lib/presentation/decorators/chat_list/title_text_style_decorator/title_text_style_decorator.dart:62]:
```suggestion
class MuteChatListTitleTextStyleDecorator implements TextStyleDecorator {
```

File: lib/presentation/decorators/chat_list/subtitle_text_style_decorator/subtitle_text_style_decorator.dart [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+  TextStyle textStyle(Room room) {
+    return _interfaceTextStyleComponent.textStyle(room);
+  }
+
+  @override
+  InterfaceChatListSubtitleTextStyleComponent get interfaceTextStyleComponent =>
+      _interfaceTextStyleComponent;
+}
+
+class SubtitleReadTextStyleDecorator
```
Comment (hoangdat) [inline lib/presentation/decorators/chat_list/subtitle_text_style_decorator/subtitle_text_style_decorator.dart:29]:
```suggestion
class ReadChatListSubtitleTextStyleDecorator
```

File: lib/presentation/decorators/chat_list/subtitle_text_style_decorator/subtitle_text_style_decorator.dart [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+    implements InterfaceChatListSubtitleTextStyleComponent {
+  @override
+  TextStyle textStyle(Room room) {
+    return LinagoraTextStyle.material().bodyMedium3.copyWith(
+          color: LinagoraSysColors.material().onSurface,
+        );
+  }
+}
+
+class SubtitleUnreadTextStyleDecorator implements TextStyleDecorator {
```
Comment (hoangdat) [inline lib/presentation/decorators/chat_list/subtitle_text_style_decorator/subtitle_text_style_decorator.dart:39]: idem

File: lib/presentation/decorators/chat_list/subtitle_text_style_decorator/subtitle_text_style_decorator.dart [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+      return _interfaceTextStyleComponent.textStyle(room);
+    }
+  }
+
+  @override
+  InterfaceChatListSubtitleTextStyleComponent get interfaceTextStyleComponent =>
+      _interfaceTextStyleComponent;
+}
+
+class SubtitleMuteTextStyleDecorator implements TextStyleDecorator {
```
Comment (hoangdat) [inline lib/presentation/decorators/chat_list/subtitle_text_style_decorator/subtitle_text_style_decorator.dart:64]: idem

File: lib/pages/forward/selectable_chat_list_item.dart [hunk 18]
```diff
@@ -1,97 +0,0 @@
...
-import 'package:fluffychat/pages/forward/forward_item_style.dart';
-import 'package:fluffychat/pages/forward/selectable_chat_list_item_style.dart';
-import 'package:fluffychat/utils/matrix_sdk_extensions/matrix_locals.dart';
-import 'package:fluffychat/widgets/avatar/avatar.dart';
-import 'package:flutter/material.dart';
-import 'package:flutter_gen/gen_l10n/l10n.dart';
-import 'package:linagora_design_flutter/colors/linagora_sys_colors.dart';
-import 'package:matrix/matrix.dart';
-
-class SelectableChatListItem extends StatelessWidget {
```
Comment (hoangdat) [inline lib/pages/forward/selectable_chat_list_item.dart:11]: how to remove it? if you want, you can use `decorator` for `selectable` also

Comment (nqhhdev) [reply lib/pages/forward/selectable_chat_list_item.dart:11]: This class is not used anywhere, I will delete it

File: lib/pages/chat_list/chat_list_item_title.dart [hunk 19]
```diff
@@ -71,17 +68,30 @@ class ChatListItemTitle extends StatelessWidget with ChatListItemMixin {
...
-                  TextStyle(
-                    letterSpacing: 0.5,
-                    color: unread
-                        ? Theme.of(context).colorScheme.onSurface
-                        : LinagoraRefColors.material().neutral[50],
-                  ),
+          padding: ChatListItemTitleStyle.paddingLeftIcon,
+          child: Row(
+            children: [
+              if (typingText.isEmpty &&
```
Comment (drminh2807) [inline lib/pages/chat_list/chat_list_item_title.dart:74]: move it to mixin

**Comment by hoangdat** [comment]
- please demo `share file` and `forward`

**Review by hoangdat — APPROVED** [review]

**Review by hoangdat — APPROVED** [review]

**Review by drminh2807 — APPROVED** [review]

**Review by sherlockvn — APPROVED** [review]

**Pull request merged**